CODING_STYLE.md

```diff
@@ -0,0 +1,166 @@
+// See https://nodejs.org/api/ for a list of Node.js core modules
+var EventEmitter = require('events').EventEmitter
+var util = require('util')
+
+// In the 2nd block, require all modules listed in package.json
+var async = require('async')
+var lodash = require('lodash')
+
+// in the 3rd block, require all modules using relative paths
+var helpers = require('./utils/helpers')
+var otherMethod = require('./other-method')
+
+function myMethod () {
+  // code here
+}
+```
+
+### Avoid "this" and object-oriented coding styles.
+
+Do this
+
+```js
+function MyApi (options) {
+  var state = {
+    foo: options.foo
+  }
+  return {
+    doSomething: doSomething.bind(null, state)
+  }
+}
+
+function doSomething (state) {
+  return state.foo ? 'foo!' : 'bar'
+}
+```
+
+Instead of
+
+```js
+function MyApi (options) {
+  this.foo = options.foo
+}
+
+MyApi.prototype.doSomething = function () {
+  return this.foo ? 'foo!' : 'bar'
+}
+```
+
+The [bind method](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/bind)
+allows for [partially applied functions](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/bind#Partially_applied_functions_%28currying%29), that way we can pass internal state between
+different methods without exposing in the public API. At the same time we can
+easily test the different methods in isolation by setting the internal state to
+what ever context we want to test with.
+
+### Folder Structure
+
+In the root, have
+
+- `package.json`
+- `.gitignore` (should at least list node_modules)
+- `README.md`
+- `LICENSE` (Apache License Version 2.0)
+
+In most cases you will have `index.js` file which is listed in `package.json`
+as the `"main"` property.
+
+If you want to split up logic into separate files, move them into a `lib/` folder.
+Put reusable, state-less helper methods into `lib/utils/`
+
+For tests, create a `test/` folder. If your module becomes a bit more complex,
+split up the tests in `test/unit` and `test/integration/`. All files that contain
+tests should end with `-test.js`.
+
+### Misc
+
+- Prefer [lodash](https://lodash.com) over [underscore](http://underscorejs.org "Underscore.js").
+
+
+## Client
+
+### Testing
+
+Client code should be tested using [tape](https://www.npmjs.com/package/tape).
+The reason we use tape is its support for [browserify](https://www.npmjs.com/package/browserify).
+
+### Libraries with sub-modules that can be required individually, like lodash
+
+For client-side JavaScript code, it is important to limit the amount of code
+that is downloaded to the client to the code that is actually needed. The
+[loadash](https://lodash.com) library is a collection of utilities that are
+useful individually and in combination.
+
+For example, if you want to use the `merge` function of lodash, require it like
+this:
+
+```javascript
+var merge = require('lodash/merge')
```

Comment:
Do we prefer to require individual `lodash` modules through their file paths in the main package, or use the individual module packages, like [`lodash.merge`](https://www.npmjs.com/package/lodash.merge) which usually means less code being downloaded during `npm install`?